src/tools/codegen-unit-test.ts

```diff
@@ -0,0 +1,137 @@
+/**
+ * @license
+ * Copyright (c) 2019 Google Inc. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import fs from 'fs';
+import {Manifest} from '../runtime/manifest.js';
+import {Flags} from '../runtime/flags.js';
+
+type Test = {
+  name: string;
+  options: object;
+  input: string;
+  results: string[];
+};
+
+/**
+ * A base class for unit tests of codegen.
+ *
+ * To use this class one has to do 2 things:
+ * - Write a .cgtest in goldens/ directory that specifies a set of
+ *   inputs and expected outputs for codegen.
+ * - Add a CodegenUnitTest to a test suite (e.g. kotlin-codegen-test-suite.ts)
+ *   with the code that should run for each input.
+ *
+ * The main benefit of using this approach is programatic updating of
+ * expectations in .cgtest files via ./tools/sigh updateCodegenUnitTests
+ *
+ * .cgtest files have one or more tests defined as following:
+ *
+ * [NAME]
+ * name of your test here
+ * [OPTS] // <- this is optional
+ * {"extraParam": 42} // JSON Object with extra arguments
+ * [RESULTS]
+ * expected output goes here
+ * [NEXT] // <- this is optional, only use for multiple outputs
+ * another expected output
+ * [END]
+ */
+export abstract class CodegenUnitTest {
+
+  constructor(
+    readonly title: string,
+    readonly inputFileName: string,
+  ) {}
+
+  get inputFilePath() {
+    return `./src/tools/tests/goldens/${this.inputFileName}`;
```

Comment:
How will this map on to G3? With the current glob rules, will it "just work"?

Reply:
It seems like this framework could be generally useful for codegen tests -- i.e. when we switch to ktpoet. WDYT?